Source/EasyNetQ/Consumer/ConsumerDispatcher.cs

```diff
@@ -8,18 +8,36 @@ namespace EasyNetQ.Consumer
     public class ConsumerDispatcher : IConsumerDispatcher
     {
         private readonly ILog logger = LogProvider.For<ConsumerDispatcher>();
-        private readonly BlockingCollection<Action> queue;
-        private bool disposed;
+        private readonly CancellationTokenSource cancellation = new CancellationTokenSource();
+        private readonly BlockingCollection<Action> durableActions = new BlockingCollection<Action>();
+        private readonly BlockingCollection<Action> transientActions = new BlockingCollection<Action>();
 
         public ConsumerDispatcher(ConnectionConfiguration configuration)
         {
             Preconditions.CheckNotNull(configuration, "configuration");
 
-            queue = new BlockingCollection<Action>();
 
             var thread = new Thread(_ =>
             {
-                while (!disposed && queue.TryTake(out var action, -1))
+                var blockingCollections = new[] {durableActions, transientActions};
+                while (!cancellation.IsCancellationRequested)
```

Comment:
I guess you have the two while loops as examples ? i prefer  while (!cancellation.IsCancellationRequested) :) but i do not really like the BlockingCollection<Action>.TakeFromAny :(

Reply:
>I guess you have the two while loops as examples? 

Nope, I have split loop to two for simplicity of the code: 1) First one processes actions from queues before the cancellation is requested, uses blocking methods and respects CancellationToken. 2) Second one processes actions after cancellation, uses nonblocking methods and completely ignores CancellationToken. 

> but i do not really like the BlockingCollection.TakeFromAny :(

Why?

Reply:
@Pliner Why BlockingCollection(s) again?! Previous solution was quite good, just change SemaphoreSlim to ManualResetEventSlim. Previous solution was simpler.

Reply:
The previous approach was quite complex because we should manage correctly two queues, semaphore, cancellation token and implement blocking behaviour on top of it.

I tried to fix it without replacement Semaphore to MRE(actually, a replacement was just an idea and the code should be fixable even with Semaphore) and it was merely complex exercise. So, the result is very similar to the idea with two blocking queues.
 
Anyway, I couldn't find any issues with the solution. Please, let me know if you will find them.

PS Probably MRE is fine, but, again, it's quite tricky code. Contributions are welcome, so if you have an alternative, feel free to provide it.

Reply:
> The previous approach was quite complex because we should manage correctly two queues, semaphore, cancellation token and implement blocking behaviour on top of them.

Well, the current approach is no simpler as it controls synchronization in the same _quite complex_ way.

> 1) First one processes actions from queues before the cancellation is requested, uses blocking methods and respects CancellationToken. 2) Second one processes actions after the cancellation, uses nonblocking methods and completely ignores CancellationToken.

The previous solution:
1. Used only non-blocking methods to take actions from queues.
2. Had only one loop, that is, the code was not duplicated.

In general, I can’t say that the current solution is better than the previous one (adjusted for ManualResetEventSlim instead Semaphore). One low-level API is replaced by another low-level API with a side effect in duplicate code. Probably I should not say that code duplication in any sense smells.

Reply:
Agree with @sungam3r it has become rather complex

Reply:
@Rotvig Do you mind to revert 0b137240248628b60153b74dd27c01838765ea0b and change semaphore to MRES? @Pliner probably we will get the simplest and most correct of all solutions.

Reply:
Not at all :) @Pliner  will you revert it ? :)

Reply:
Done.

Reply:
@Pliner that look really great :) good idea with the semaphore :)

Reply:
@Rotvig Glad to hear. 

Anyway, it has several bugs mentioned ^ and their fixing is not in my plans, so feel free to fix it if you want :wink:

PS The only acceptable solution for now (from my point of view) was mentioned in this commit(https://github.com/EasyNetQ/EasyNetQ/pull/1021/commits/0b137240248628b60153b74dd27c01838765ea0b).

Reply:
Hmm okay, maybe we have to use https://github.com/EasyNetQ/EasyNetQ/commit/0b137240248628b60153b74dd27c01838765ea0b then

Reply:
@Rotvig Just change semaphore to ManualResetEventSlim. That is all.

Reply:
I would not be so sure @sungam3r :wink:

Reply:
Guys can we get this PR merge at some point ? XD @sungam3r  @Pliner which solution would you prefer ? 

I would prefer the one i made with AutoResetEvent :)  
But i am completely okay with the two loops and blocking queues as well :)

Reply:
> Guys can we get this PR merge at some point ? XD @sungam3r @Pliner which solution would you prefer ?
> 
> I would prefer the one i made with AutoResetEvent :)
> But i am completely okay with the two loops and blocking queues as well :)

Cool. Let's merge https://github.com/EasyNetQ/EasyNetQ/commit/0b137240248628b60153b74dd27c01838765ea0b then and see how it will behave.

BTW I'm not so sure that this stuff will be necessary with `IAsyncBasicConsumer`, will investigate how to integrate it to develop branch.

Reply:
> which solution would you prefer ?

I have already explained everything above. I have not heard counterarguments about the proposal to use MRES.

Reply:
Let's move all conversation related to the issue to slack.